clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java

```diff
@@ -41,12 +43,20 @@ public ProduceRequestResult() {
      * Mark this request as complete and unblock any threads waiting on its completion.
      * @param topicPartition The topic and partition to which this record set was sent was sent
      * @param baseOffset The base offset assigned to the record
+     * @param responseTimestamp The timestamp returned by the broker
      * @param error The error that occurred if there was one, or null.
      */
-    public void done(TopicPartition topicPartition, long baseOffset, RuntimeException error) {
+    public void set(TopicPartition topicPartition, long baseOffset, long responseTimestamp, RuntimeException error) {
         this.topicPartition = topicPartition;
         this.baseOffset = baseOffset;
+        this.responseTimestamp = responseTimestamp;
         this.error = error;
+    }
+
+    /**
+     * Mark this request as complete and unblock any threads waiting on its completion.
+     */
+    public void done() {
```

Comment:
Yes, it is a requirement to call `set` before calling `done`. However, I thought about this some more and perhaps it's OK to combine the two methods into one _if_ we move the `done` before the callbacks are invoked. I didn't want to do this because it's a slight change in behaviour, but it seems to be allowed by the specification and I can't think of a reasonable case where this could have a negative impact.